apps/server/src/shared/domain/rules/submission.rule.spec.ts

```diff
@@ -0,0 +1,279 @@
+import { MikroORM } from '@mikro-orm/core';
+import { Test, TestingModule } from '@nestjs/testing';
+import {
+	courseFactory,
+	courseGroupFactory,
+	roleFactory,
+	setupEntities,
+	submissionFactory,
+	taskFactory,
+	userFactory,
+} from '@shared/testing';
+import { CourseGroupRule, CourseRule, LessonRule, SubmissionRule, TaskRule } from '.';
+import { User } from '../entity';
+import { Permission, RoleName } from '../interface';
+import { Actions } from './actions.enum';
+
+describe('SubmissionRule', () => {
+	let orm: MikroORM;
+	let submissionRule: SubmissionRule;
+	let student: User;
+	let student2: User;
+	let teacher: User;
+	const permissionA = 'a' as Permission;
+	const permissionB = 'b' as Permission;
+	const permissionC = 'c' as Permission;
+
+	beforeAll(async () => {
+		orm = await setupEntities();
+
+		const module: TestingModule = await Test.createTestingModule({
+			providers: [TaskRule, CourseRule, LessonRule, CourseGroupRule, SubmissionRule],
+		}).compile();
+
+		submissionRule = await module.get(SubmissionRule);
+
+		const studentRole = roleFactory.build({ permissions: [permissionA, permissionB], name: RoleName.STUDENT });
+		student = userFactory.build({ roles: [studentRole] });
+		student2 = userFactory.build({ roles: [studentRole] });
+		const teacherRole = roleFactory.build({ permissions: [permissionC], name: RoleName.TEACHER });
+		teacher = userFactory.build({ roles: [teacherRole] });
+	});
+
+	afterAll(async () => {
+		await orm.close();
+	});
+
+	describe('isApplicable', () => {
+		it('should return true if entity is instance of Submission', () => {
+			const task = taskFactory.build();
+			const submission = submissionFactory.build({ task, student });
+
+			const result = submissionRule.isApplicable(student, submission);
+
+			expect(result).toBe(true);
+		});
+
+		it('should return false if entity is not instance of Submission', () => {
+			const task = taskFactory.build();
+
+			const result = submissionRule.isApplicable(student, task);
+
+			expect(result).toBe(false);
+		});
+	});
+
+	describe('hasPermission', () => {
+		describe('given user is teacher', () => {
```

Comment:
It is documentated and the consens in the project that we use when for it. 
If we want to change it, it must be first go over the Arc round and after it, we can change it on any place.

We have also the decision to use setup fr the when describes that hold the preparation to this area like factorys stuff, mocks and so on.

Reply:
I replaced "given" with "when".